File: xlsx.go
package main

import (
	"fmt"
	"strings"

	"github.com/xuri/excelize/v2"
)

const xlsxTemplateSheetName = "产品集导入模板"

// ExportXlsxTemplate writes a styled xlsx template to the given path.
func (a *App) ExportXlsxTemplate(path string) ApiResult[bool] {
	if strings.TrimSpace(path) == "" {
		return Err[bool]("路径不能为空")
	}

	f := excelize.NewFile()
	defer f.Close()

	// Rename default sheet.
	index, err := f.GetSheetIndex("Sheet1")
	if err != nil {
		return Err[bool](err.Error())
	}
	_ = f.SetSheetName(f.GetSheetName(index), xlsxTemplateSheetName)

	headers := []string{"产品集名称"}
	example := []string{"示例产品集"}

	// Header style: bold text, blue background.
	headerStyleID, err := f.NewStyle(&excelize.Style{
		Font:      &excelize.Font{Bold: true, Color: "#FFFFFF"},
		Fill:      excelize.Fill{Type: "pattern", Color: []string{"#3B82F6"}, Pattern: 1},
		Alignment: &excelize.Alignment{Horizontal: "center", Vertical: "center"},
	})
	if err != nil {
		return Err[bool](err.Error())
	}

	// Example row style: gray italic text.
	exampleStyleID, err := f.NewStyle(&excelize.Style{
		Font: &excelize.Font{Italic: true, Color: "#6B7280"},
		Fill: excelize.Fill{Type: "pattern", Color: []string{"#F3F4F6"}, Pattern: 1},
	})
	if err != nil {
		return Err[bool](err.Error())
	}

	// Write headers.
	for col, h := range headers {
		cell, _ := excelize.CoordinatesToCellName(col+1, 1)
		_ = f.SetCellValue(xlsxTemplateSheetName, cell, h)
	}
	_ = f.SetCellStyle(xlsxTemplateSheetName, "A1", "A1", headerStyleID)

	// Write example row.
	for col, v := range example {
		cell, _ := excelize.CoordinatesToCellName(col+1, 2)
		_ = f.SetCellValue(xlsxTemplateSheetName, cell, v)
	}
	_ = f.SetCellStyle(xlsxTemplateSheetName, "A2", "A2", exampleStyleID)

	// Add instructions sheet.
	instSheet := "填写说明"
	if _, err := f.NewSheet(instSheet); err != nil {
		return Err[bool](err.Error())
	}
	_ = f.SetCellValue(instSheet, "A1", "产品集导入模板使用说明")
	_ = f.SetCellStyle(instSheet, "A1", "A1", headerStyleID)
	_ = f.SetCellValue(instSheet, "A3", "1. 在“产品集导入模板”工作表中填写数据，从第 3 行开始。")
	_ = f.SetCellValue(instSheet, "A4", "2. 每行对应一个产品集，导入后仅创建产品集。")
	_ = f.SetCellValue(instSheet, "A5", "3. 第 2 行为示例数据，填写前请删除或覆盖。")
	_ = f.SetCellValue(instSheet, "A6", "4. 产品集名称为必填项，不可为空。")
	_ = f.SetColWidth(instSheet, "A", "A", 80)

	// Set column widths on data sheet.
	_ = f.SetColWidth(xlsxTemplateSheetName, "A", "A", 24)

	if err := f.SaveAs(path); err != nil {
		return Err[bool](err.Error())
	}
	return Ok(true)
}

// ImportProductSetsFromXlsx reads an xlsx file and batch creates product sets.
func (a *App) ImportProductSetsFromXlsx(path string) ApiResult[[]ProductSetInfo] {
	ws := a.currentWorkspacePath()
	if ws == "" {
		return Err[[]ProductSetInfo]("未打开工作区")
	}
	path = strings.TrimSpace(path)
	if path == "" {
		return Err[[]ProductSetInfo]("路径不能为空")
	}

	f, err := excelize.OpenFile(path)
	if err != nil {
		return Err[[]ProductSetInfo](err.Error())
	}
	defer f.Close()

	// Use the first sheet.
	sheet := f.GetSheetName(0)
	if sheet == "" {
		return Err[[]ProductSetInfo]("无法读取工作表")
	}

	rows, err := f.GetRows(sheet)
	if err != nil {
		return Err[[]ProductSetInfo](err.Error())
	}
	if len(rows) <= 1 {
		return Ok([]ProductSetInfo{})
	}

	// Collect unique product set names.
	seen := make(map[string]bool)
	created := make([]ProductSetInfo, 0)
	for i, row := range rows[1:] {
		lineNo := i + 2 // 1-based line number in spreadsheet (header is row 1)
		if len(row) == 0 || allEmpty(row) {
			continue
		}
		setName := strings.TrimSpace(getCell(row, 0))
		if setName == "" {
			return Err[[]ProductSetInfo](fmt.Sprintf("第 %d 行产品集名称不能为空", lineNo))
		}
		if seen[setName] {
			continue
		}
		seen[setName] = true

		result := a.ProductSetCreate(ProductSetCreateRequest{Name: setName})
		if !result.Success {
			return Err[[]ProductSetInfo](result.Error)
		}
		created = append(created, result.Data)
	}

	return Ok(created)
}

func getCell(row []string, index int) string {
	if index < 0 || index >= len(row) {
		return ""
	}
	return row[index]
}

func allEmpty(row []string) bool {
	for _, c := range row {
		if strings.TrimSpace(c) != "" {
			return false
		}
	}
	return true
}

File: src/main/core/xlsx.ts
/**
 * XLSX 模板导出/批量导入（对照原 Go xlsx.go，exceljs 替代 excelize）
 * 纯 TS 业务层：不 import electron，可在 node 环境直接测试。
 * 性能：exceljs 延迟加载（动态 import），避免主进程启动加载其依赖链。
 */
import ExcelJS from 'exceljs'
import { WorkspaceService, ProductSetInfo } from './workspace'

const TEMPLATE_SHEET = '产品集导入模板'
const INSTRUCTIONS_SHEET = '填写说明'

function allEmpty(row: ExcelJS.Row): boolean {
  let hasValue = false
  row.eachCell({ includeEmpty: false }, () => {
    hasValue = true
  })
  return !hasValue
}

export class XlsxService {
  constructor(private workspace: WorkspaceService) {}

  /** 导出带样式的导入模板（对照 ExportXlsxTemplate） */
  async exportTemplate(filePath: string): Promise<void> {
    if (!filePath.trim()) throw new Error('路径不能为空')
    const XLSX = ExcelJS
    const wb = new XLSX.Workbook()
    const ws = wb.worksheets[0] ?? wb.addWorksheet(TEMPLATE_SHEET)
    ws.name = TEMPLATE_SHEET

    // 表头：粗体白字蓝底居中
    const header = ws.getCell('A1')
    header.value = '产品集名称'
    header.font = { bold: true, color: { argb: 'FFFFFFFF' } }
    header.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF3B82F6' } }
    header.alignment = { horizontal: 'center', vertical: 'middle' }

    // 示例行：灰色斜体 + 浅灰底
    const example = ws.getCell('A2')
    example.value = '示例产品集'
    example.font = { italic: true, color: { argb: 'FF6B7280' } }
    example.fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FFF3F4F6' } }

    ws.getColumn(1).width = 24

    // 填写说明 sheet
    const inst = wb.addWorksheet(INSTRUCTIONS_SHEET)
    inst.getCell('A1').value = '产品集导入模板使用说明'
    inst.getCell('A1').font = { bold: true, color: { argb: 'FFFFFFFF' } }
    inst.getCell('A1').fill = { type: 'pattern', pattern: 'solid', fgColor: { argb: 'FF3B82F6' } }
    inst.getCell('A3').value = '1. 在“产品集导入模板”工作表中填写数据，从第 3 行开始。'
    inst.getCell('A4').value = '2. 每行对应一个产品集，导入后仅创建产品集。'
    inst.getCell('A5').value = '3. 第 2 行为示例数据，填写前请删除或覆盖。'
    inst.getCell('A6').value = '4. 产品集名称为必填项，不可为空。'
    inst.getColumn(1).width = 80

    await wb.xlsx.writeFile(filePath)
  }

  /** 从 xlsx 批量创建产品集（对照 ImportProductSetsFromXlsx） */
  async importProductSets(filePath: string): Promise<ProductSetInfo[]> {
    const ws0 = this.workspace.currentWorkspacePath()
    if (!ws0) throw new Error('未打开工作区')
    if (!filePath.trim()) throw new Error('路径不能为空')

    const XLSX = ExcelJS
    const wb = new XLSX.Workbook()
    await wb.xlsx.readFile(filePath)
    const sheet = wb.worksheets[0]
    if (!sheet) throw new Error('无法读取工作表')

    const created: ProductSetInfo[] = []
    const seen = new Set<string>()
    for (let r = 2; r <= sheet.rowCount; r++) {
      const row = sheet.getRow(r)
      if (allEmpty(row)) continue
      const name = String(row.getCell(1).value ?? '').trim()
      if (!name) throw new Error(`第 ${r} 行产品集名称不能为空`)
      if (seen.has(name)) continue
      seen.add(name)
      created.push(await this.workspace.productSetCreate({ name }))
    }
    return created
  }
}
